src/gz/bufread.rs

```diff
@@ -338,34 +401,50 @@ impl<'a, T: Read> Read for Buffer<'a, T> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         if self.buf_cur == self.buf_max {
             let len = self.reader.read(buf)?;
-            self.buf.extend_from_slice(&buf[..len]);
+            self.part.buf.extend_from_slice(&buf[..len]);
             Ok(len)
         } else {
-            let len = (&self.buf[self.buf_cur..self.buf_max]).read(buf)?;
+            let len = (&self.part.buf[self.buf_cur..self.buf_max]).read(buf)?;
             self.buf_cur += len;
             Ok(len)
         }
     }
 }
 
+impl<'a, T> Buffer<'a, T>
+where
+    T: std::io::Read,
+{
+    // If we manage to read all the bytes, we reset the buffer
+    fn read_once(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.buf_cur = 0;
+        self.read_exact(buf)?;
+        let rlen = buf.len();
+        self.crc.update(buf);
+        self.part.buf.truncate(0);
+        self.buf_max = 0;
```

Comment:
Trying to read over this some more, I don't think I understand why this method needs to exist. I don't think that the manipulation of the fields of the structure here (e.g. the updates of `buf_{cur,max}` and the `truncate` are right. That seems like it's discarding data which might otherwise be succesfully read by subsequent calls to `read_once`. 

This seems like callers of this function should be calling `read_exact` and then on success updating the crc manually for the header? Other than that I'm not sure why this function is needed?

Reply:
The problem with `read_exact` is that we do not reset the buffer, leading to the quadratic complexity as we keep parsing all over and again.

What I want is
`read_exact` but buffer it there is not enough data (that was already ok), and clear the buffer if it succeeded

For example, the reader first provides 7 bytes, `read_gz_header_part` buffers these 7 bytes and returns `ErrorKind::WouldBlock`

Then a second call to `read_gz_header_part`, the reader provides 4 more bytes
There is a successful read of the 10 bytes header
As we have flag `FEXTRA`, we try to read a `u16` but fail with `ErrorKind::WouldBlock`
The problem with the current code is that we have a 11 byte buffer instead of a 1 byte one for later calls to `read_gz_header_part`

By the way, I see a problem with my code that arc should belong to `GzHeaderPartial` as it lives longer than `Buffer`

Reply:
Nah yeah I understand the quadratic part, and I forgot that this is where the fix for that is going. To fix the quadratic bits, though, I think that the `.bytes()` iterator can be replaced when reading the nul-terminated string with custom logic to empty the buffer. Basically when reading the C strings for the various fields I think that we should consume the bytes and not buffer them in the general intermediate buffer, but rather buffer them in the location where the field is read from (basically store them directly where they're gonna go).

Reply:
>  Basically when reading the C strings for the various fields I think that we should consume the bytes and not buffer them in the general intermediate buffer, but rather buffer them in the location where the field is read from (basically store them directly where they're gonna go).

I did that, so `Buffer::read` looks at GzHeaderParsingState to decide where to buffer the data.

The point I do not like about this is that the logic is now split between `read_gz_header_part` and `Buffer::read`, ie we do not see in `read_gz_header_part` that we are filling the filename